src/sage/algebras/quatalg/quaternion_algebra.py

```diff
@@ -2870,30 +2938,142 @@ def is_equivalent(self, J, B=10) -> bool:
+        r"""
+        Checks whether ``self`` and ``J`` are equivalent as right ideals.
+        Requires the underlying quaternion algebra to be definite.
+
+        INPUT:
+
+        - ``J`` -- a fractional quaternion right ideal with same order as ``self``
+
+        - ``B`` -- a bound to compute and compare theta series before
+          doing the full equivalence test
+
+        - ``certificate`` -- if ``True`` returns an element alpha such that alpha*J=I
+
+        OUTPUT: bool, or (bool, alpha) if ``certificate`` is ``True``
         """
-        # shorthand: let I be self
-        if not isinstance(self, QuaternionFractionalIdeal_rational):
+        if not isinstance(J, QuaternionFractionalIdeal_rational):
+            if certificate:
+                return False, None
             return False
 
         if self.right_order() != J.right_order():
-            raise ValueError("self and J must be right ideals")
+            raise ValueError("self and J must be right ideals over the same order")
+
+        if not self.quaternion_algebra().is_definite():
+            raise NotImplementedError('equivalence test of ideals not implemented'
+                                    'for indefinite quaternion algebras')
 
-        # Just test theta series first.  If the theta series are
-        # different, the ideals are definitely not equivalent.
+        # Just test theta series first; if the theta series are
+        # different, the ideals are definitely not equivalent
         if B > 0 and self.theta_series_vector(B) != J.theta_series_vector(B):
+            if certificate:
+                return False, None
             return False
 
         # The theta series are the same, so perhaps the ideals are
-        # equivalent.  We use Prop 1.18 of [Pizer, 1980] to decide.
+        # equivalent.  We use Prop 1.18 of [Piz1980]_ to decide.
         # 1. Compute I * Jbar
-        # see Prop. 1.17 in Pizer.  Note that we use IJbar instead of
-        # JbarI since we work with right ideals
+        # Note that we use I*Jbar instead of Jbar*I since we work with right ideals
         IJbar = self.multiply_by_conjugate(J)
 
-        # 2. Determine if there is alpha in K such
-        #    that N(alpha) = N(I)*N(J) as explained by Pizer.
+        # 2. Determine if there is alpha in K such that
+        # N(alpha) = N(I)*N(J) as explained by Pizer
         c = IJbar.theta_series_vector(2)[1]
-        return c != 0
+        if c == 0:
+            if certificate:
+                return False, None
+            return False
+
+        if not certificate:
+            return True
+
+        # Use Pari's qfminim with flag = 1 to find alpha as above
+        _,v = IJbar.quadratic_form().__pari__().qfminim(None, None, 1)
+        alpha = sum(ZZ(c)*g for c,g in zip(v, IJbar.basis()))
+
+        # As explained by Pizer, we now only need to rescale alpha by 1/N(J)
+        return True, alpha * (1/J.norm())
+
+    def is_principal(self, certificate=False):
```

Comment:
This method duplicates quite some code from the equivalence tests. Perhaps the equivalence tests could/should just end up calling `.is_principal()` on `IJbar` instead of repeating all the steps on that special case?